### ISSUE TITLE
2070 RTX: LossTensor is inf or nan (redux)

### ISSUE THREAD

**Issue by tbenst** [DeepLabCut/DeepLabCut#837]
**Describe the bug**
On one of my datasets, I kept hitting https://github.com/DeepLabCut/DeepLabCut/issues/62. By displaying each iteration, I got a sense for how editing my .csv files could delay this issue. I discovered that DLC really does not like frames with no targets, i.e. [this csv file](https://gist.github.com/tbenst/dbe9616091354bc0a53ad46e51716183) reliably blows up the loss and crashes training. In this case it's fine to remove the lines with `,,,,,`, but sometimes there are frames where no targets are visible and DLC should be able to handle these cases, perhaps via gradient clipping.

**To Fix (for stray search engine refugees)**
1) open all csv in `labeled_data` folder in editor of choice, eg `code */*NAME.csv` for vscode
2) remove all labeled frames with all blanks or mostly blank
3) `dlc.convertcsv2h5(config_yaml, userfeedback=False)`
3) check labels
4) create training dataset

edit: did not fully resolve issue :/. Everything looks normal and then suddenly....loss explodes!

```
iteration: 190 loss: 0.0150 lr: 0.005
iteration: 191 loss: 0.0388 lr: 0.005
iteration: 192 loss: 0.0145 lr: 0.005
iteration: 193 loss: 0.0358 lr: 0.005
iteration: 194 loss: 0.0156 lr: 0.005
iteration: 195 loss: 0.0271 lr: 0.005
iteration: 196 loss: 0.0243 lr: 0.005
iteration: 197 loss: 0.0267 lr: 0.005
iteration: 198 loss: 0.0184 lr: 0.005
iteration: 199 loss: 0.0218 lr: 0.005
iteration: 200 loss: 0.0253 lr: 0.005
iteration: 201 loss: 0.0262 lr: 0.005
iteration: 202 loss: 0.0228 lr: 0.005
iteration: 203 loss: 0.0174 lr: 0.005
iteration: 204 loss: 0.0160 lr: 0.005
iteration: 205 loss: 0.0232 lr: 0.005
iteration: 206 loss: 0.0233 lr: 0.005
iteration: 207 loss: 0.0168 lr: 0.005
iteration: 208 loss: 0.0419 lr: 0.005
iteration: 209 loss: 0.0168 lr: 0.005
iteration: 210 loss: 0.0221 lr: 0.005
iteration: 211 loss: 0.0254 lr: 0.005
iteration: 212 loss: 0.0427 lr: 0.005
iteration: 213 loss: 0.0269 lr: 0.005
iteration: 214 loss: 0.0377 lr: 0.005
iteration: 215 loss: 0.0174 lr: 0.005
iteration: 216 loss: 0.0324 lr: 0.005
iteration: 217 loss: 0.0294 lr: 0.005
iteration: 218 loss: 0.0302 lr: 0.005
iteration: 219 loss: 0.0383 lr: 0.005
iteration: 220 loss: 0.0433 lr: 0.005
iteration: 221 loss: 0.0237 lr: 0.005
iteration: 222 loss: 0.0150 lr: 0.005
iteration: 223 loss: 0.0210 lr: 0.005
iteration: 224 loss: 0.0358 lr: 0.005
iteration: 225 loss: 0.0256 lr: 0.005
iteration: 226 loss: 0.0221 lr: 0.005
iteration: 227 loss: 0.0205 lr: 0.005
iteration: 228 loss: 0.0204 lr: 0.005
iteration: 229 loss: 0.0358 lr: 0.005
iteration: 230 loss: 2696.8628 lr: 0.005
2020-08-05 10:45:41.525535: E tensorflow/core/kernels/check_numerics_op.cc:185] abnormal_detected_host @0x7ff448a0cc00 = {1, 0} LossTensor is inf or nan
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/client/session.py", line 1322, in _do_call
    return fn(*args)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/client/session.py", line 1307, in _run_fn
    options, feed_dict, fetch_list, target_list, run_metadata)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/client/session.py", line 1409, in _call_tf_sessionrun
    run_metadata)
tensorflow.python.framework.errors_impl.InvalidArgumentError: LossTensor is inf or nan : Tensor had NaN values
	 [[Node: train_op/CheckNumerics = CheckNumerics[T=DT_FLOAT, _class=["loc:@train_op/control_dependency"], message="LossTensor is inf or nan", _device="/job:localhost/replica:0/task:0/device:GPU:0"](add)]]
	 [[Node: train_op/control_dependency/_1651 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device_incarnation=1, tensor_name="edge_10156_train_op/control_dependency", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"]()]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python3.6/dist-packages/deeplabcut/pose_estimation_tensorflow/training.py", line 134, in train_network
    raise e
  File "/usr/local/lib/python3.6/dist-packages/deeplabcut/pose_estimation_tensorflow/training.py", line 132, in train_network
    train(str(poseconfigfile),displayiters,saveiters,maxiters,max_to_keep=max_snapshots_to_keep,keepdeconvweights=keepdeconvweights,allow_growth=allow_growth) #pass on path and file name for pose_cfg.yaml!
  File "/usr/local/lib/python3.6/dist-packages/deeplabcut/pose_estimation_tensorflow/train.py", line 190, in train
    feed_dict={learning_rate: current_lr})
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/client/session.py", line 900, in run
    run_metadata_ptr)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/client/session.py", line 1135, in _run
    feed_dict_tensor, options, run_metadata)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/client/session.py", line 1316, in _do_run
    run_metadata)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/client/session.py", line 1335, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.InvalidArgumentError: LossTensor is inf or nan : Tensor had NaN values
	 [[Node: train_op/CheckNumerics = CheckNumerics[T=DT_FLOAT, _class=["loc:@train_op/control_dependency"], message="LossTensor is inf or nan", _device="/job:localhost/replica:0/task:0/device:GPU:0"](add)]]
	 [[Node: train_op/control_dependency/_1651 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device_incarnation=1, tensor_name="edge_10156_train_op/control_dependency", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"]()]]

Caused by op 'train_op/CheckNumerics', defined at:
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python3.6/dist-packages/deeplabcut/pose_estimation_tensorflow/training.py", line 132, in train_network
    train(str(poseconfigfile),displayiters,saveiters,maxiters,max_to_keep=max_snapshots_to_keep,keepdeconvweights=keepdeconvweights,allow_growth=allow_growth) #pass on path and file name for pose_cfg.yaml!
  File "/usr/local/lib/python3.6/dist-packages/deeplabcut/pose_estimation_tensorflow/train.py", line 151, in train
    learning_rate, train_op = get_optimizer(total_loss, cfg)
  File "/usr/local/lib/python3.6/dist-packages/deeplabcut/pose_estimation_tensorflow/train.py", line 102, in get_optimizer
    train_op = slim.learning.create_train_op(loss_op, optimizer)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/contrib/slim/python/slim/learning.py", line 439, in create_train_op
    check_numerics=check_numerics)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/contrib/training/python/training/training.py", line 464, in create_train_op
    'LossTensor is inf or nan')
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/gen_array_ops.py", line 737, in check_numerics
    "CheckNumerics", tensor=tensor, message=message, name=name)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/ops.py", line 3392, in create_op
    op_def=op_def)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/ops.py", line 1718, in __init__
    self._traceback = self._graph._extract_stack()  # pylint: disable=protected-access

InvalidArgumentError (see above for traceback): LossTensor is inf or nan : Tensor had NaN values
	 [[Node: train_op/CheckNumerics = CheckNumerics[T=DT_FLOAT, _class=["loc:@train_op/control_dependency"], message="LossTensor is inf or nan", _device="/job:localhost/replica:0/task:0/device:GPU:0"](add)]]
	 [[Node: train_op/control_dependency/_1651 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device_incarnation=1, tensor_name="edge_10156_train_op/control_dependency", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"]()]]
```

edit2: I am running using https://github.com/DeepLabCut/Docker4DeepLabCut2.0/blob/testing/Dockerfile

**Comment by AlexEMG** [comment]
I see! Yes, this is due to the conversion I suppose. In lieu of labels there will be a nan, which is then skipped https://github.com/DeepLabCut/DeepLabCut/blob/master/deeplabcut/generate_training_dataset/trainingsetmanipulation.py#L810

Can you please verify why it passes this without a label?

**Comment by tbenst** [comment]
@AlexEMG my diagnosis was wrong unfortunately. I just created a new project, extracted frames for one video, labeled 3 of them, check labels, create training dataset, then train (resnet152), and I got the exact same error, no conversion involved.

...might this be an issue with the dockerfile? A bit at a loss. Using latest `testing`.

**Comment by tbenst** [comment]
Ok, I found two new bugs that may/may not be related.

## tensorpack error
I  just tried tensorpack and got a different error (didn't even make it to training):

```
Traceback (most recent call last):
  File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/local/lib/python3.6/dist-packages/tensorpack/dataflow/parallel.py", line 312, in run
    dp = next(itr)
  File "/usr/local/lib/python3.6/dist-packages/tensorpack/dataflow/parallel.py", line 50, in _repeat_iter
    yield from get_itr()
  File "/usr/local/lib/python3.6/dist-packages/tensorpack/dataflow/common.py", line 315, in __iter__
    for dp in self.ds:
  File "/usr/local/lib/python3.6/dist-packages/tensorpack/dataflow/common.py", line 316, in __iter__
    ret = self.func(copy(dp))  # shallow copy the list
  File "/usr/local/lib/python3.6/dist-packages/deeplabcut/pose_estimation_tensorflow/dataset/pose_dataset_tensorpack.py", line 236, in augment
    tfm = aug.get_transform(img)
  File "/usr/local/lib/python3.6/dist-packages/tensorpack/dataflow/imgaug/meta.py", line 35, in get_transform
    return self.aug.get_transform(img)
  File "/usr/local/lib/python3.6/dist-packages/tensorpack/dataflow/imgaug/geometry.py", line 220, in get_transform
    transform_matrix = cv2.getRotationMatrix2D(tuple(center), deg, scale)
TypeError: Argument 'angle' can not be safely parsed to 'double'
```

This appeared in each worker.

## updating frame label does not propagate to training dataset

For my new test with five labeled frames, I only labeled 3/4 markers for frame 1, and 4/4 for frames 2-5. I then did check_frames and create_dataset. when I look at `open-field-v2-mattwright-2020-08-05/training-datasets/iteration-0/UnaugmentedDataSet_open-field-v2Aug5/CollectedData_mattwright.csv`, as expected I see https://gist.github.com/tbenst/b48f8dc1b579fa18c75975f253afb76a. I then I go back and update the first frame to have the fourth label, and then run check labels and create training dataset again. Although the labeled frame is different and shows up in GUI, the training-datasets csv is unchanged! I also tried deleting iteration-0 folder, but on recreation the csv file still has a blank fourth marker.

**Comment by MMathisLab** [comment]
(note that the csv is never used, only the h5 file, so perhaps double check this)

Also, can you please fill out the template as requested; what version of DLC, what platform (conda, docker, etc)

**Comment by tbenst** [comment]
I had problems with https://github.com/DeepLabCut/Docker4DeepLabCut2.0/blob/testing/Dockerfile, so I created a new Dockerfile for deeplabcut that is more consistent with installation instructions: https://github.com/tbenst/Docker4DeepLabCut2.0/blob/tyler/Dockerfile

Curiously, the computer used *does* matter. When I run this dockerfile on NixOS 20.03 host it works fine, but running on Ubuntu 20.04 host leads to the aforementioned "LossTensor is inf or nan" problem. Natively running on Ubuntu 18.04 also works (haven't tried docker yet).

On NixOS, I'm using 440.82 as the driver, and on Ubuntu, I'm using `nvidia-driver-440-server` package, which is 440.82. Ubuntu 20.04 seems to add the choice between nvidia-driver-440 and nvidia-driver-440-server, not sure what the difference is or if that matters but only the latter is marked at "tested".

Curiously, after I get this error in the docker container, `nvidia-smi` tends to become incredibly slow, taking 15+ seconds to report GPUs. I just noticed that both GPU Fan and Pwr now show `ERR!`. Thus, it seems like a hardware problem, or at least a driver problem. As such I think I've solved this issue's main problem, although the bugs in https://github.com/DeepLabCut/DeepLabCut/issues/837#issuecomment-669163281 are still valid.

Edit: Ok, I have no idea what is going on. The fan/power ERR was a red herring. My best guess now is seems like there's something problematic about the latest nvidia-docker and old versions of CUDA, but curious if anyone else can reproduce issues with deeplabcut on Ubuntu 20.04

Edit 2: it's a hardware problem! swapping out the 2070 RTX for a Titan RTX completely resolved the issue.
